src/EFCore.PG/Query/Internal/NpgsqlSetOperationTypeResolutionCompensatingExpressionVisitor.cs

```diff
@@ -0,0 +1,152 @@
+using Microsoft.EntityFrameworkCore.Query.Internal;
+
+namespace Npgsql.EntityFrameworkCore.PostgreSQL.Query.Internal;
+
+public class NpgsqlSetOperationTypeResolutionCompensatingExpressionVisitor : ExpressionVisitor
+{
+    private State _state;
+
+    protected override Expression VisitExtension(Expression extensionExpression)
+        => extensionExpression switch
+        {
+            ShapedQueryExpression shapedQueryExpression
+                => shapedQueryExpression.Update(
+                    Visit(shapedQueryExpression.QueryExpression),
+                    Visit(shapedQueryExpression.ShaperExpression)),
+            SetOperationBase setOperationExpression => VisitSetOperation(setOperationExpression),
+            SelectExpression selectExpression => VisitSelect(selectExpression),
+            TpcTablesExpression tpcTablesExpression => VisitTpcTablesExpression(tpcTablesExpression),
+            _ => base.VisitExtension(extensionExpression)
+        };
+
+    private Expression VisitSetOperation(SetOperationBase setOperationExpression)
+    {
+        switch (_state)
+        {
+            case State.Nothing:
+                _state = State.InSingleSetOperation;
+                var visited = base.VisitExtension(setOperationExpression);
+                _state = State.Nothing;
+                return visited;
+
+            case State.InSingleSetOperation:
+                _state = State.InNestedSetOperation;
+                visited = base.VisitExtension(setOperationExpression);
+                _state = State.InSingleSetOperation;
+                return visited;
+
+            default:
+                return base.VisitExtension(setOperationExpression);
+        }
+    }
+
+#pragma warning disable EF1001
```

Comment:
@smitpatel as discussed offline, am merging with this temporary hack which accesses TpcTablesExpression - will clean this up later based on what you decide.